src/webhooks.ts

```diff
@@ -1,109 +1,74 @@
 import * as url from 'url'
 
-import { PackageInitParams } from './types'
-import { sign, verify } from './util/signature'
-import { getPublicKey } from './util/publicKey'
+import { sign } from './util/signature'
 import { parseSignatureHeader } from './util/parser'
+import { isSignatureHeaderValid, hasEpochExpired } from './util/webhooks'
+import { MissingSecretKeyError, WebhookAuthenticateError } from './errors'
 
-/**
- * Helper function to construct the basestring and verify the signature of an
- * incoming request
- * @param uri String
- * @param submissionId MongoDB submission ObjectId
- * @param formId MongoDB submission ObjectId
- * @param epoch Number of milliseconds since Jan 1, 1970
- * @param signature base64 encoded signature
- * @param webhookPublicKey base64 webhook public key
- */
-function verifySignature(
-  uri: string,
-  submissionId: string,
-  formId: string,
-  epoch: number,
-  signature: string,
-  webhookPublicKey: string
-) {
-  const baseString = `${url.parse(uri).href}.${submissionId}.${formId}.${epoch}`
-  return verify(baseString, signature, webhookPublicKey)
-}
+export default class Webhooks {
+  publicKey: string
+  secretKey?: string
 
-/**
- * Helper function to verify that the epoch submitted is recent.
- * Prevents against replay attacks.
- * @param epoch The number of milliseconds since Jan 1, 1979
- * @param expiry Duration of expiry. The default is 5 minutes.
- */
-function verifyEpoch(epoch: number, expiry: number = 300000) {
-  const difference = Date.now() - epoch
-  return difference > 0 && difference < expiry
-}
+  constructor({
+    publicKey,
+    secretKey,
+  }: {
+    publicKey: string
+    secretKey?: string
+  }) {
+    this.publicKey = publicKey
+    this.secretKey = secretKey
+  }
 
-/**
- * Higher order function that injects the webhook public key for authentication
- * @param webhookPublicKey The FormSG webhook public key
- */
-function authenticate(webhookPublicKey: string) {
   /**
    * Injects the webhook public key for authentication
    * @param header X-FormSG-Signature header
    * @param uri The endpoint that FormSG is POSTing to
+   * @returns true if the header is verified
    * @throws {Error} If the signature or uri cannot be verified
    */
-  function _internalAuthenticate(header: string, uri: string) {
+  authenticate = (header: string, uri: string) => {
     // Parse the header
+    const signatureHeader = parseSignatureHeader(header)
     const {
       v1: signature,
-      t,
+      t: epoch,
       s: submissionId,
       f: formId,
-    } = parseSignatureHeader(header)
-    const epoch = Number(t)
+    } = signatureHeader
 
     if (!epoch || !signature || !submissionId || !formId) {
-      throw new Error('X-FormSG-Signature header is invalid')
+      throw new WebhookAuthenticateError('X-FormSG-Signature header is invalid')
     }
```

Comment:
I think these checks should be part of `isSignatureHeaderValid()` no?

Reply:
Yea, agreed. Moved into validation function.